docs/proposals/kubemark.md

```diff
@@ -0,0 +1,147 @@
+
+# Kubemark proposal
+
+## Goal of this document
+
+This document describes a design of Kubemark - a system that allows performance testing of a Kubernetes cluster. It describes the assumption, high level design and discusses possible solutions for lower-level problems. It is supposed to be a starting point for more detailed discussion.
+
+## Current state and objective
+
+Currently performance testing happens on ‘live’ clusters of up to 100 Nodes. It takes quite a while to start such cluster or to push updates to all Nodes, and it uses quite a
+lot of resources. At this scale the amount of wasted time and used resources is still acceptable. In the next quarter or two we’re targeting 1000 Node cluster, which will push
+it way beyond ‘acceptable’ level. Additionally we want to enable people without many resources to run scalability tests on bigger clusters
+than they can afford at given time. Having an ability to cheaply run scalability tests will enable us to run some set of them on "normal"
+test clusters, which in turn would mean ability to run them on every PR.
+
+This means that we need a system that will allow for realistic performance testing on (much) smaller number of “real” machines. First assumption we make is that Nodes are
+independent, i.e. number of existing Nodes do not impact a performance of a single Node. This is not entirely true, as number of Nodes can increase latency of various
+components on Master machine, which in turn may increase latency of Node operations, but we’re not interested in measuring this effect here. In turn we want to measure how
+number of Nodes and the load imposed by Node daemons affects the performance of Master components.
+
+## Kubemark architecture overview
+
+The high-level idea behind Kubemark is to write library that allows running artificial "Hollow" Nodes that will be able to simulate a behavior of real Kubelet and KubeProxy in a single, lightweight
+binary. Hollow components will need to correctly respond to Controllers (via API server), and preferably, in the fullness of time, be able to ‘replay’
+previously recorded real traffic (this is out of scope for initial version). In the initial version we only want them to be able to fool Master
+components and put some configurable (in what way TBD) load on them. To teach Hollow components replaying recorded traffic they will need
+to store data specifying when given Pod/Container should die (e.g. observed lifetime). Such data can be extracted e.g. from etcd Raft logs,
+or it can be reconstructed from Events.
+
+When we have Hollow Node ready, we’ll be able to test performance of Master Components by creating a real Master Node, with API server, Controllers, etcd and whatnot, and
+create number of Hollow Nodes that will register to the running Master.
+
+For the initial version we will teach Master components to use port number to identify Kubelet/KubeProxy. This will allow running those components on non-default ports, and in
```

Comment:
> What I'm proposing is a host cluster, which is a totally ordinary kubernetes cluster. On it we would run a "hollow" cluster, by compiling node components into pods that run a HollowNode, and grouping master components into a pod that acts in place of a master VM.

OK, that sounds like it matches what I wrote above: a set of real nodes, a set of real master components to manage the real nodes, a set of hollow nodes running as pods on the real nodes, and a second set of master components that are managing the hollow cluster.

> But this is equivalent to saying that one can't run workloads on kubernetes, because kubernetes itself consumes too much resources? 

No, just saying it may be tricky to factor out the performance impact of the infrastructure that is managing the hollow cluster, since it is running on the same physical machines and network as the hollow cluster. I agree that having two sets of master components solves one potential performance crosstalk problem ,but there is still the one I mentioned.